src/core/algorithms/od/fastod/partitions/complex_stripped_partition.cpp

```diff
@@ -93,6 +93,115 @@ void ComplexStrippedPartition::ToStrippedPartition() {
     should_be_converted_to_sp_ = false;
 }
 
+template <bool Ascending>
+bool ComplexStrippedPartition::Swap(model::ColumnIndex left, model::ColumnIndex right) const {
+    size_t const group_count = is_stripped_partition_ ? sp_begins_->size() : rb_begins_->size();
+
+    for (size_t begin_pointer = 0; begin_pointer < group_count - 1; begin_pointer++) {
+        size_t const group_begin = is_stripped_partition_ ? (*sp_begins_)[begin_pointer]
+                                                          : (*rb_begins_)[begin_pointer];
+
+        size_t const group_end = is_stripped_partition_ ? (*sp_begins_)[begin_pointer + 1]
+                                                        : (*rb_begins_)[begin_pointer + 1];
+
+        std::vector<std::pair<int, int>> values;
+
+        if (is_stripped_partition_) {
+            values.reserve(group_end - group_begin);
+
+            for (size_t i = group_begin; i < group_end; ++i) {
+                size_t const index = (*sp_indexes_)[i];
+
+                values.emplace_back(data_->GetValue(index, left), data_->GetValue(index, right));
+            }
+        } else {
+            for (size_t i = group_begin; i < group_end; ++i) {
+                DataFrame::Range const range = (*rb_indexes_)[i];
+
+                for (size_t j = range.first; j <= range.second; ++j) {
+                    values.emplace_back(data_->GetValue(j, left), data_->GetValue(j, right));
+                }
+            }
+        }
+
+        if constexpr (Ascending) {
+            std::sort(values.begin(), values.end(),
+                      [](auto const& p1, auto const& p2) { return p1.first < p2.first; });
+        } else {
+            std::sort(values.begin(), values.end(),
+                      [](auto const& p1, auto const& p2) { return p2.first < p1.first; });
+        }
+
+        size_t prev_group_max_index = 0;
+        size_t current_group_max_index = 0;
+        bool is_first_group = true;
+
+        for (size_t i = 0; i < values.size(); i++) {
+            auto const& [first, second] = values[i];
+
+            if (i != 0 && values[i - 1].first != first) {
+                is_first_group = false;
+                prev_group_max_index = current_group_max_index;
+                current_group_max_index = i;
+            } else if (values[current_group_max_index].second <= second) {
+                current_group_max_index = i;
+            }
+
+            if (!is_first_group && values[prev_group_max_index].second > second) {
+                return true;
+            }
+        }
```

Comment:
This loop constantly checking for `i != 0`.
Moreover, the first iteration of this loop will always look like
```cpp
if (values[0].second <= values[0].second) //current_group_max_index is zero
    current_group_max_index = 0;
```
since `i == 0` and `!is_first_group` is false.

So we can just start from `i = 1` and remove that check